### PR TITLE
Add contributing guidelines

### PR DESCRIPTION
**Pull request by TylerJang27**
The `contributing.md` file includes an overview of expectations for adding new linters and actions. Additional testing details are reserved for the testing guidelines in `tests/readme.md`.

### CODE REVIEW

File: contributing.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      ref: v0.0.8
+```
+
+Plugins can also be loaded from local paths, as below:
+
+```yaml
+plugins:
+  sources:
+    - id: trunk
+      local: /path/to/repo/root
```
Comment (det) [inline contributing.md:31]:
```suggestion
      local: </path/to/repo/root>
```

Perhaps this is better to make it clear it is a placeholder.

File: contributing.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+   ```
+
+3. Modify plugin.yaml to define your new linter. Refer to other linters in this repository as
+   examples.
+4. Making sure the plugin in `.trunk/trunk.yaml` is pointing to your local repository, run
+   `trunk check enable <my-linter>` to enable your linter, and run `trunk check` to verify that the
+   configuration is valid and that you get desired diagnostics. Running `trunk check --verbose` can
+   help provide greater insights when debugging.
+5. Add a few simple test cases to `my_linter.test.ts` to exercise your linter and generate
+   snapshots. Refer to [Testing Guidelines](tests/readme.md) for more information on writing and
```
Comment (det) [inline contributing.md:88]: Do we need to document `linter-test-helper` somewhere?

Comment (TylerJang27) [reply contributing.md:88]: Not sure on this point. I made a note about files should be created for you, but I think it's more of a nice-to-have and not really a robust tool we want to advertise too overtly? Open to thoughts.

Comment (TylerJang27) [reply contributing.md:88]: Post-discussion. No need to document now, until more robust

File: tests/readme.md [hunk 3]
```diff
@@ -141,7 +138,7 @@ PRs will run 4 types of tests:
 
 ### Debugging
 
-Occasionally, tests may take a while to run due to installing of linter dependencies. Subsequent
+Occasionally, tests may take a while to run due to installation of linter dependencies. Subsequent
```
Comment (sxlijin) [inline tests/readme.md:141]:
```suggestion
Tests normally complete in <5 minutes. They may take up to ~15 minutes or so if the dependency cache is empty (e.g. linters need to be downloaded and installed to run the linter tests).
```

adjust numbers as appropriate.

File: contributing.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+If you need a linter or action definition from this repo that hasn't been released yet, you can
+change the `ref` field of a plugin source to be any `git` ref. However, we recommend sticking to our
+official tagged releases unless absolutely necessary.
+
+## Linters
+
+If you want to add a new linter integration with trunk, refer to our docs on
+[custom linters](https://docs.trunk.io/docs/check-custom-linters) and
+[custom parsers](https://docs.trunk.io/docs/custom-parsers), and work through the following:
```
Comment (sxlijin) [inline contributing.md:64]:
```suggestion
To add a new linter:
```

"If you want to" and "with trunk" are both redundant here: if a user is in these docs, they want to do so; and these docs are all in the context of trunk.

File: contributing.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+     │ plugin.yaml
+     │ my_linter.test.ts
+     │ readme.md (optional)
+     │ my-config.json (optional)
+     └─test_data/
+        └─basic.in.py (with appropriate extension)
+   ```
+
+3. Modify plugin.yaml to define your new linter. Refer to other linters in this repository as
+   examples.
```
Comment (sxlijin) [inline contributing.md:82]:
```suggestion
3. Add your linter definition to `plugin.yaml` (consult the docs for [custom linters] and [custom parsers] to understand how it should be defined)
```

File: contributing.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+   `trunk check enable <my-linter>` to enable your linter, and run `trunk check` to verify that the
+   configuration is valid and that you get desired diagnostics. Running `trunk check --verbose` can
+   help provide greater insights when debugging.
+5. Add a few simple test cases to `my_linter.test.ts` to exercise your linter and generate
+   snapshots. Refer to [Testing Guidelines](tests/readme.md) for more information on writing and
+   running tests.
+6. Run `trunk check` to lint your changes.
+7. Commit your entire linter subdirectory and create a
+   [Pull Request](https://github.com/trunk-io/plugins/compare). We will review your changes as soon
+   as we can.
```
Comment (sxlijin) [inline contributing.md:93]:
```suggestion
7. Open a PR with your changes!
```

Anyone contributing should know all of this :)

File: contributing.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+7. Run trunk check to lint your changes.
+8. Commit your entire action subdirectory and create a Pull Request. We will review your changes as
+   soon as we can.
+
+## Guidelines
+
+Please follow the guidelines below when contributing:
+
+- After defining a new linter or action, please add it to the repository's root
+  [readme.md](readme.md).
```
Comment (sxlijin) [inline contributing.md:129]:
```suggestion
- After defining a new linter or action, please add it to [`readme.md`](readme.md).
```

File: contributing.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+- [trunk](https://docs.trunk.io/docs/compatibility) supports running on most versions of Linux and
+  macOS. If your linter only runs on certain OSs, refer to the example of
+  [stringslint](linters/stringslint/stringslint.test.ts).
```
Comment (sxlijin) [inline contributing.md:132]: This should be in the linters section.

File: contributing.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+Run `trunk run toggle-local` to quickly toggle this setting.
+
+Adding a plugin source lets users run `trunk check enable` or `trunk actions enable` with linters
+and actions defined in that plugin. For more information, see our
+[docs](https://docs.trunk.io/docs/plugins).
+
+If you want to request a new feature or get help with a problem, please see our
+[Features Page](https://features.trunk.io/) or our [Slack](https://slack.trunk.io/).
+
+## Releases
```
Comment (sxlijin) [inline contributing.md:43]: This section still feels a bit detail-heavy and doesn't feel like it naturally answers a user's question.

Stepping back, think about the journey/flow for someone contributing to plugins:

1. uses a linter `magicwand`
2. discovers `trunk`
3. opens PR to trunk-io/plugins with `magicwand` linter definition
4. wants to use `magicwand` in their own repo

IMO, the purpose of this section is to explain to the user what the process for going from step 3 to step 4 is, so we'd want to explain:

* trunk-io/plugins has a release process that is independent of PRs; that is, we don't live at HEAD but rather pin the version of a plugin (explain that we do this for the same reason we pin linter versions)
  * also explain what `ref` field this is (e.g. "in the above example, we're pointing at v0.0.8 of plugins")
* once a plugins release containing the desired change is created, a user can pick those changes up using `trunk upgrade`

----

In a similar vein, there may be other user flows that you have in mind for someone reading this doc. Remember: always keep your audience in mind when writing.

(related/useful: https://nick.groenen.me/posts/the-4-types-of-technical-documentation/)

**Review by sxlijin — COMMENTED** [review]
much better!

File: contributing.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+```
+
+Run `trunk run toggle-local` to quickly toggle this setting in this repo.
+
+Adding a plugin source lets users run `trunk check enable` or `trunk actions enable` with linters
+and actions defined in that plugin. For more information, see our
+[docs](https://docs.trunk.io/docs/plugins).
+
+If you want to request a new feature or need help, please see our
+[Features Page](https://features.trunk.io/) or our [Slack](https://slack.trunk.io/).
```
Comment (sxlijin) [inline contributing.md:41]:
```suggestion
If you have questions, please [stop by our community Slack]; and if you have a feature request, you can [file it here].
```

File: contributing.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+## Releases
+
+Plugins are released on a fairly regular cadence that is independent of PRs. Users will pick up
```
Comment (sxlijin) [inline contributing.md:45]:
```suggestion
`trunk-io/plugins` is released on a fairly regular cadence that is independent of PRs. Users will pick up
```

File: contributing.md [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+these configuration changes by running `trunk upgrade` to automatically update their plugin version
+to the latest release.
+
+```yaml
+plugins:
+  sources:
+    - id: trunk
+      uri: https://github.com/trunk-io/plugins
+      ref: v0.0.8 # will change to v0.0.9
```
Comment (sxlijin) [inline contributing.md:54]:
```suggestion
      ref: v0.0.8  # will change to v0.0.9 on next "trunk upgrade"
```

**Review by sxlijin — APPROVED** [review]

**Pull request merged**